src/dmd/statementsem.d

```diff
@@ -2510,7 +2510,7 @@ else
                 ss.condition.type = ss.condition.type.constOf();
                 break;
             }
-            ss.condition = integralPromotions(ss.condition, sc);
+
```

Comment:
The error messages produced indicate that these checks are backwards, and it is a mistake to move the integralPromotions from here.

Reply:
In the first message I explained that promotion is kept only to prevent a problem with the code produced by the DMD backend. For example you can test this code when the promotion is completely removed:

```d
string test(char a)
{
    version (fix) asm
    {
        xor EAX, EAX;
    }
    switch (a)  // mov AL, byte ptr[...], but RAX is garbage, then cmp EAX result are randomly wrong
    {
        case 'a': return "a";
        case 'b': return "b";
        default: assert(false); // will take this path because of the garbages in EAX.
    }
}

void main(string[] args)
{
    assert(test('a') == "a");
    assert(test('b') == "b");
} 
```
Without the BE issue the promotion would be removed. Why do you want to promote a `byte` to an `int` if the goal is to test the different values that a `byte` can have ? 

Impossible cases are accepted because of the promotion that happened before the semantic check of the case and that is based on the promoted target type, i.e `ss.condition.type`.

Reply:
You are indeed correct that:

"The case expressions must all evaluate to a constant value or array, or a runtime initialized const or immutable variable of integral type. They must be implicitly convertible to the type of the switch Expression."
https://dlang.org/spec/statement.html#switch-statement

I had that backwards.

Reply:
yeah but nothing says in the specs that the type of the condition has to be promoted. Of course if it gets promoted before processing the cases that then it's obvious why the errors are not issued...

So I'm not sure if I understand. Do you agree on the fact that there's a bug ?

The bug to me is like if `enum byte a = ubyte(-1);` would be accepted just because where are in a switch, while everywhere else that fails.